<html>
<head>
<title>{{servername|title}}</title>

<style type='text/css'>
 .header {text-align:center; font-size:2em; font-weight:bolder; font-family:verdana,arial,sans-serif}
 .strong {color:blue; font-weight:bold; font-size:1.2em}
 .red {color:red;}
 fieldset {border-style:solid; border-width:medium; border-color:gray; margin:20px 0 8px 0px;}
 body {font-family:verdana, arial, sans-serif; font-size:0.8em}
 .divisions {border: 1px solid ; padding: 5px 5px 5px 5px; background: #EFEFEF; margin: 20px 0 8px 0px;}
 .divisionTitle {position: relative; top: -12px; border: 1px solid; background: #e6f2ff; width: 180px; margin: 0 0 0 0; font-weight: bold; font-size: 1em; padding: 0 0 0 2px; text-align: center;}

input[type="text"] {
	padding: 0px 0px 0px 0px;
	height:1.4em;
	vertical-align:middle;
	margin-bottom:0.2em;
	margin-top:0.2;
}

.btn {
        background:#050;
        color:white;
        font-family:'trebuchet ms',verdana,sans-serif;
	font-size:1.2em ;
        font-weight:bold;
        margin-left: 30px;
        margin-right: 10px;
}

.btc {
        background:#8B0000;
	color:white ;
	font-size:1.2em ;
        font-family:'trebuchet ms',verdana,sans-serif;
        font-weight:bold;
}


table.reference{ 
	border-collapse:collapse;
	width:100%; }
table.reference tr:nth-child(odd)	{background-color:#F6F4F0;}
table.reference tr:nth-child(even)	{background-color:#ffffff;}
table.reference tr.fixzebra		{background-color:#F6F4F0;}
table.reference th{ 
	color:#ffffff;
	background-color:#555555;
	border:1px solid #555555;
	padding:3px;
	vertical-align:top;
        font-size:0.5em ;
	text-align:left; } 
table.reference th a:link,table.reference th a:visited{ color:#ffffff; } 
table.reference th a:hover,table.reference th a:active{ color:#EE872A; }
table.reference td{ 
        font-size:0.5em ;
	border:1px solid #d4d4d4;
	padding:5px;
	padding-top:2px;
	padding-bottom:2px;
	vertical-align:top; }
table.reference td.example_code { vertical-align:bottom; }

</style>

<SCRIPT type="text/javascript">

//utility function to check if the data string starts with input string.
function startsWith( data, input ) {
    return data.substring(0, input.length) === input ;
}

//utility function to check if the given string or number is integer or not.
function isInt(n) 
{
    return n != "" && !isNaN(n) && Math.round(n) == n;
}

/* 1.1.2: Fixed a bug where trailing . in e-mail address was passing
           (the bug is actually in the weak regexp engine of the browser; I
           simplified the regexps to make it work).
  1.1.1: Removed restriction that countries must be preceded by a domain,
           so abc@host.uk is now legal.  However, there's still the
           restriction that an address must end in a two or three letter
           word.
    1.1: Rewrote most of the function to conform more closely to RFC 822.
    1.0: Original  */


function emailCheck (emailStr) {

	if ( (emailStr == null) || (emailStr=="") ){
	       alert("Email address is required to submit a job.") ;
	       return false ;
	}
	/* The following pattern is used to check if the entered e-mail address
	  fits the user@domain format.  It also is used to separate the username
	  from the domain. */
	var emailPat=/^(.+)@(.+)$/ ;
	/* The following string represents the pattern for matching all special
	  characters.  We don't want to allow special characters in the address.
	  These characters include ( ) < > @ , ; : \ " . [ ]    */
	var specialChars="\\(\\)<>@,;:\\\\\\\"\\.\\[\\]" ;
	/* The following string represents the range of characters allowed in a
	  username or domainname.  It really states which chars aren't allowed. */
	var validChars="\[^\\s" + specialChars + "\]" ;
	/* The following pattern applies if the "user" is a quoted string (in
	  which case, there are no rules about which characters are allowed
	  and which aren't; anything goes).  E.g. "jiminy cricket"@disney.com
	  is a legal e-mail address. */
	var quotedUser="(\"[^\"]*\")" ;
	/* The following pattern applies for domains that are IP addresses,
	  rather than symbolic names.  E.g. joe@[123.124.233.4] is a legal
	  e-mail address. NOTE: The square brackets are required. */
	var ipDomainPat=/^\[(\d{1,3})\.(\d{1,3})\.(\d{1,3})\.(\d{1,3})\]$/ ;
	/* The following string represents an atom (basically a series of
	  non-special characters.) */
	var atom=validChars + '+' ;
	/* The following string represents one word in the typical username.
	  For example, in john.doe@somewhere.com, john and doe are words.
	  Basically, a word is either an atom or quoted string. */
	var word="(" + atom + "|" + quotedUser + ")" ;
        // The following pattern describes the structure of the user
	var userPat=new RegExp("^" + word + "(\\." + word + ")*$") ;
	/* The following pattern describes the structure of a normal symbolic
	  domain, as opposed to ipDomainPat, shown above. */
	var domainPat=new RegExp("^" + atom + "(\\." + atom +")*$") ;


	/* Finally, let's start trying to figure out if the supplied address is
	  valid. */

	/* Begin with the coarse pattern to simply break up user@domain into
	  different pieces that are easy to analyze. */
	var matchArray=emailStr.match(emailPat) ;
	if (matchArray==null) {
	 /* Too many/few @'s or something; basically, this address doesn't
	    even fit the general mould of a valid e-mail address. */
	       alert("Email address seems incorrect (check @ and .'s)") ;
	       return false ;
	}
	var user=matchArray[1] ;
	var domain=matchArray[2] ;

	// See if "user" is valid
	if (user.match(userPat)==null) {
	   // user is not valid
	   alert("The username doesn't seem to be valid.") ;
	   return false ;
	}

	/* if the e-mail address is at an IP address (as opposed to a symbolic
	  host name) make sure the IP address is valid. */
	var IPArray=domain.match(ipDomainPat) ;
	if (IPArray!=null) {
	   // this is an IP address
		 for (var i=1;i<=4;i++) {
		   if (IPArray[i]>255) {
		       alert("Destination IP address is invalid!") ;
		       return false ;
		   }
	   }
	   return true ;
	}

	// Domain is symbolic name
	var domainArray=domain.match(domainPat) ;
	if (domainArray==null) {
	       alert("The domain name doesn't seem to be valid.") ;
	   return false ;
	}

	/* domain name seems valid, but now make sure that it ends in a
	  three-letter word (like com, edu, gov) or a two-letter word,
	  representing country (uk, nl), and that there's a hostname preceding
	  the domain or country. */

	/* Now we need to break up the domain to get a count of how many atoms
	  it consists of. */
	var atomPat=new RegExp(atom,"g") ;
	var domArr=domain.match(atomPat) ;
	var len=domArr.length ;
	if (domArr[domArr.length-1].length<2 || domArr[domArr.length-1].length>3) { 
	  // the address must end in a two letter or three letter word.
	  alert("The address must end in a three-letter domain, or two letter country.") ;
	  return false ;
	}

	// Make sure there's a host name preceding the domain.
	if (len<2) {
	  var errStr="This address is missing a hostname!" ;
	  alert(errStr) ;
	  return false ;
	}

	// If we've gotten this far, everything's valid!
	return true;
}

function get_column_number( str ) {
    return str.split( "\t" ).length ;
}

function all_same_columns( lines ) {
    var n = lines[0].split("\t").length ; 
    for ( var i=1 ; i < lines.length ; i++ ) {
	if ( n != lines[i].split("\t").length ) {
            return false ;
        }
    }
    return true ;
}

function split_lines( str ) {
    return str.split(/\r\n|[\n\v\f\r\x85\u2028\u2029]/) ;
}


function get_table_preview( lines ) {
    //number of columns
    var n = get_column_number( lines[0] ) ;


    ////
    //table header
    ////
    var table = "<tr>" ;
    for ( var i=1 ; i <= n ; i++ ) {
        table += "<th id=col"+i+">Column " + i + "</th>" ;
    }
    table += "</tr>" ;
    
    ////
    //table body
    ////
 
    //The first row is seperated to add identifiers to each cell.
    var row = "<tr>" ;
    var columns = lines[i].split( "\t" ) ;
    for ( var j = 0 ; j < columns.length ; j++ ) {
        row += "<td id=col"+(j+1)+"_data>" + columns[j] + "</td>" ;
    }
    table += row + "</tr>" ; 

    //rest of the rows
    for ( var i = 1 ; i < lines.length ; i++ ) {
        row = "<tr>" ;
        columns = lines[i].split( "\t" ) ;
        for ( var j = 0 ; j < columns.length ; j++ ) {
            row += "<td>" + columns[j] + "</td>" ;
        }
        table += row + "</tr>" ; 
        if (i >= 10 ) { 
            break ; 
        }
    }

    return "<strong>File Content Preview</strong><br><table class=reference>" + table + "</table>" ;
}


function get_select_options( ncols ) {
    var text = '<option></option>' ;
    for ( var i=1 ; i <= ncols ; i++ ) {
        text += '<option>col'+ i +'</option>' ;
    }
    return text ;
}


function reset_column_headers( ncols ) {
    for ( var i=1 ; i <= ncols ; i++ ) {
        var hname = 'col' + i ;
        document.getElementById(hname).innerHTML='Column '+ i
    }
    document.getElementById('customSelectDiv').innerHTML=''
}


function set_header( id, text ) {
    document.getElementById( id ).innerHTML = text ;
}

function set_narrowpeak(ncols) {
    //Need to reset column headers before putting info
    //just in case other functions messed the header before this call
    reset_column_headers(ncols) ;

    set_header( 'col1', 'Chromosome Name' ) ;
    set_header( 'col2', 'Peak Start' ) ;
    set_header( 'col3', 'Peak End' ) ;
    set_header( 'col10', 'Peak Summit' ) ;
}

function set_summitcol(el, ncols) {
    for ( var i = 1 ; i <= ncols ; i++ ) {
        var ele = document.getElementById('col'+i) ;
        
        if ( ele && ele.innerHTML == 'Peak Summit' ) {
            ele.innerHTML = 'Column ' + i ;
        }
    }
    var summitcol = el.options[ el.selectedIndex ].text
    if (summitcol) {
        document.getElementById( summitcol ).innerHTML = 'Peak Summit' ;
    }
    else {
        alert( 'Peak summit column should be selected. If your file does not contain peak summit information select nomalization by peak center.' ) ;
    }
}

function get_summit_selector( ncols ) {
    var text = '<strong>Peak Summit Column &nbsp</strong><select id="summitcol" name="summitcol" onchange="set_summitcol(this,'+ncols+')">' ;
    text += '<option></option>' ;
    for ( var i=4 ; i<=ncols; i++ ) {
        text += '<option>col'+ i +'</option>' ;
    }
    return text + '<select>' ;
}

function set_bed( ncols ) {
    reset_column_headers(ncols) ;
    set_header( 'col1', 'Chromosome Name' ) ;
    set_header( 'col2', 'Peak Start' ) ;
    set_header( 'col3', 'Peak End' ) ;

    var normsummit = document.getElementById('normsummit') ;
    if (normsummit.checked) {
       	document.getElementById('customSelectDiv').innerHTML = get_summit_selector(ncols) ;
    }
}

function set_normtype( ncols ) {
    var narrowpeak = document.getElementById( 'narrowpeak' ).checked ;
    var bed = document.getElementById( 'bedpeak' ).checked ;
    if ( narrowpeak ) {
        set_narrowpeak(ncols) ;
        return true ;
    }
    if (bedpeak) {
        set_bed(ncols) ;
        return true ;
    }
}


function get_file_type_selector( ncols ) {
    var text = '<div class=divisions><div class=divisionTitle>Select normalization</div><input type=radio name=normtype value=summit id=normsummit onclick="set_normtype('+ncols+')">By Peak Summits</input><br>'
        text += '<input type=radio name=normtype value=center id=normcenter onclick="set_normtype('+ncols+')" >By Peak Centers</input></div>'

        text += '<div class=divisions><div class=divisionTitle>Select file format</div><input type=radio name=fileformat value=narrowpeak id=narrowpeak onclick="set_narrowpeak('+ncols+')">NarrowPeak format</input><br>' ;
        text +='<input type=radio name=fileformat value=bed id=bedpeak onclick="set_bed('+ncols+')">BED format</input><br>' ;
        text += '<div id=customSelectDiv></div>'

	/*
        text += '<br>' ;
        text += '<select id="chrcol" name="chrcol" onchange="select_file_type(this)">' ;
        text += get_select_options(ncols) ;
        text += '</select>' ;
        text += '<select id="startcol" name="startcol" onchange="select_file_type(this)">' ;
        text += get_select_options(ncols) ;
        text += '</select>' ;
        text += '<select id="endcol" name="endcol" onchange="select_file_type(this)">' ;
        text += get_select_options(ncols) ;
        text += '</select>' ;
        text += '<select id="summitcol" name="summitcol" onchange="select_file_type(this)">' ;
        text += get_select_options(ncols) ;
        text += '</select>' ;
	*/

    return text ;
}

function handleFileSelect(el) {
    // initialize the preview area
    document.getElementById("filepreview").innerHTML = '';

    var files = el.files; // FileList object saved in input file tag.

    // files is a FileList of File objects. List some properties.
    var output = [] ;
    var f = files[0] ;
    output.push('<li><strong>', escape(f.name), '</strong> (', f.type || 'n/a', ') - ',
                f.size, ' bytes, last modified: ',
                f.lastModifiedDate ? f.lastModifiedDate.toLocaleDateString() : 'n/a',
                '</li>') ;
    //document.getElementById('fileopt').innerHTML = '<ul>' + output.join('') + '</ul>';

    //read the first 10kb file for preview
    var start = 0;
    var stop = 1024*10;
    if ( stop > f.size ) {
        stop = f.size - 1 ;
    }

    var reader = new FileReader();
    reader.onloadend = function(e) {
      if (e.target.readyState == FileReader.DONE) { // DONE == 2

	//////
        //raw output preview
        //document.getElementById("filepreview").innerHTML = e.target.result;
	//////

	//////
	//output preview as a table
	var lines = split_lines(e.target.result) ;
        lines.pop() ; //remove the last line for safety...

	if( lines.length <= 1 ) {
            alert( "Too few lines detected! Are you sure the file is in correct BED or NarrowPeak formats?" ) ;
            return false ;
        }


	if ( !all_same_columns(lines) ) {
            alert("Rows have different number of columns! Are you sure the file is in BED or NarrowPeak formats?") ;
	    return false ;
        }
	var ncols = get_column_number( lines[0] ) ;
        if ( ncols < 3 ) {
            alert( "At least three columns (Chromosome Name, Peak Start and Peak End) are required to discover motifs." ) ;
            return false;
        }

	if ( lines.length > 1 ) {
            document.getElementById("filepreview").innerHTML = get_file_type_selector( ncols ) ;
            document.getElementById("filepreview").innerHTML += get_table_preview( lines ) ;
	}
	else {
            alert( "Too few lines detected! Are you sure the file is in BED or NarrowPeak formats?" ) ;
            return false ;
        }

        if (ncols == 10 ) {
            document.getElementById( "normsummit" ).checked = true ;
            document.getElementById( "narrowpeak" ).checked = true ;
            set_narrowpeak( ncols ) ;
        }
        else if (ncols == 3) {
            document.getElementById( "narrowpeak" ).disabled = true ;
            document.getElementById( "normsummit" ).disabled = true ;
            document.getElementById( "normcenter" ).checked = true ;
            document.getElementById( "bedpeak" ).checked = true ;
            set_bed( ncols ) ;
        }
	else { 
            document.getElementById( "narrowpeak" ).disabled = true ;
            document.getElementById( "normsummit" ).checked = true ;
            document.getElementById( "bedpeak" ).checked = true ;
            set_bed( ncols ) ;
	}

        //code snippet from the html5 blog example
        //document.getElementById('byte_range').textContent = 
        //    ['Read bytes: ', start + 1, ' - ', stop + 1,
        //     ' of ', file.size, ' byte file'].join('');
      }
    };

    var blob = f.slice( start, stop + 1) ;
    reader.readAsBinaryString(blob) ;
  }

function checkChrStartEndColumnValidity( chrCol, startCol, endCol ) {
    if (!isInt(document.getElementById(startCol).innerHTML)){
      	alert( 'Peak start column should contain an integer.' ) ; 
        return false ;
    }
    if (!isInt(document.getElementById(endCol).innerHTML)){
      	alert( 'Peak end column should contain an integer.' ) ; 
        return false ;
    }

    if (!startsWith(document.getElementById(chrCol).innerHTML, 'chr')){
        alert( 'Chromosome name should start from "chr"!' ) ;
        return false ;
    }

    start = parseInt( document.getElementById(startCol).innerHTML ) ;
    end = parseInt( document.getElementById(endCol).innerHTML ) ;

    if (end < start) {
        alert( 'Peak end should be bigger than peak start!' ) ;
        return false ;
    }
    
}

function checkPeakSummitColumnValidity( startCol, endCol, summitCol ) {
   if (!isInt(document.getElementById(summitCol).innerHTML)){
      alert( 'Peak summit column should contain an integer.' ) ;
      return false ;
   }

   start = parseInt( document.getElementById(startCol).innerHTML ) ;
   end = parseInt( document.getElementById(endCol).innerHTML ) ;
   summit = parseInt( document.getElementById(summitCol).innerHTML ) ;


   if ( summit < 0 ) {
      alert( 'Peak summit column value should be equal to or bigger than zero!' ) ;
      return false ;
   }
   
   if ( start + summit > end ) {
      alert( 'Summit location is outside [Start, End] range. Peak summit column value should be given as an offset from peak start column as defined in NarrowPeak format.' ) ;
      return false ;
   }
   return true ;
}


function contentCheck() {
    var selectedFile = document.getElementById( 'fileinput' ) ;
    if (selectedFile.files.length==0) {
    	alert( 'Peak file should be selected.' ) ;
    	return false ;
    }
    var norm_summit = document.getElementById('normsummit') ;
    if ( norm_summit == null ) {
        alert( 'Peak file format should be a BED or NarrowPeak format.' );
        return false ;
    }

    //format check!
    if( checkChrStartEndColumnValidity( 'col1_data', 'col2_data', 'col3_data' )  == false )
        return false ;

    var bed_peak = document.getElementById( 'bedpeak' ) ;
    if (norm_summit.checked && bed_peak.checked ) {
        var summit_col = document.getElementById( 'summitcol' ) ;
        if ( summit_col.options[ summit_col.selectedIndex ].text == '' ) {
            alert( 'Peak summit column should be selected. If your file does not have peak summit offset information, select normalization by peak center option.' ) ;
            return false ;
        }
        
        summit_data_id = summit_col.options[ summit_col.selectedIndex ].text + '_data' ;
        if ( checkPeakSummitColumnValidity( 'col2_data', 'col3_data', summit_data_id ) == false )
            return false ;
    }

    //narrow peak format
    if (bed_peak.checked == false) {
        if( checkPeakSummitColumnValidity( 'col2_data', 'col3_data', 'col10_data' ) == false )
            return false ;
    }

    return true;
}

 
</script>
</head>

<body>
<table width=100% border=0><tr><td><img src='http://www.factorbook.org/mediawiki/factordata/factorbooklogo5.png' width=100px height=100px /></td><td class='header'>{{servername|title}}</td></tr></table>
<hr noshade size=5 color=Gold>

{{servername|capitalize}} discovers <b><i>de novo</i> motifs </b> for a set of detected peaks from a ChIP-Seq experiment.<br>
Our server provides further filtering for de novo discovered motifs by measuring GC content bias and local sequence bias provided by the flanking regions.
<p>Click for the <a href={{hostparent}}/doc/flowchart.pdf target='_blank'><b>flowchart</b></a> of {{servername|title}}.<br>

Click for <a href="{{hostparent}}/doc/fmp_tutorial.html"><b>tutorial and explanation</b></a> of the server results.<br>
Click for <a href="http://www.factorbook.org"><b>ENCODE2 ChIP-seq motif analysis in factorbook.org</b></a>.
<hr noshade size=5 color=Gold>

<!--<p style="color:red; font-size:large; font-style:oblique;font-weight:bold; text-decoration:blink"> We are upgrading the server. If you see any abnormal results, retry after few minitues. </p>-->

<form name="upload_form" onSubmit="return (contentCheck() && emailCheck(this.email.value));" action='{{hostparent}}/upload' method='post' enctype='multipart/form-data'>
<center><input class='btn' name=example_search type=button onclick="window.location.href='{{hostparent}}/view/{{ exampleid }}'" value="Output example" /></center>
<fieldset width=100%>
<legend class=strong >Peak Data Input</legend>
<b>Input a query ChIP-seq peak dataset in <a href=http://genome.ucsc.edu/FAQ/FAQformat.html#format1 target='_blank'>BED format</a> or <a href=http://genome.ucsc.edu/FAQ/FAQformat.html#format12 target="_blank">ENCODE NarrowPeak format</a> to find motifs.</b> <br>Note that the peak file should be ordered by peak strength, e.g. sorted in an order for peaks with stronger signals before peaks with weaker signals. Ideally, our method works better with more than 1000 peaks detected from ChIP-Seq experiments, although it is not an absolute requirement.<br><br>

<b>Upload a peak file:</b> <input type=file name='myFile' size=30 id=fileinput onchange='handleFileSelect(this)'>
<output id=filepreview></output>
<output id=fileopt ></output>

</fieldset>

<fieldset>
<legend class=strong>Genome</legend>

<select name=db> 
	<option selected value=hg19> Human (HG19) </option>
	<option value=mm9> Mouse (MM9) </option>
</select>
<br>

</fieldset>

<fieldset> <legend class=strong>Submission </legend>
<table border=0 style="font-size:1em">
<tr><td><b>Email (required):</b></td><td> <input type='text' name='email' size=30></td></tr>
<tr><td><b>TF Name:</b></td><td> <input type='text' name='jobname' size=30></td></tr>
<tr><td><b>Data Source:</b></td><td> <input type='text' name='cell' size=30></td></tr>
<tr><td><b>Experimental Condition:</b></td><td> <input type='text' name='treatment' size=30></td></tr>
<tr><td><b>Lab Name:</b></td> <td><input type='text' name='lab' size=30></td></tr>
</table>

<center><input type='submit' value='Submit' class=btn> <input class=btc type='reset'></center>
</fieldset></form>
<br>
<hr noshade size=5 color=Gold>
{% if reference %}
    <p><b>References:</b></p>
    {{reference}}
{% endif %}
<p><i>Comments and suggestions to: <a href='mailto:{{admin_email}}'>{{admin_email}}</a></i></p>
<p><i>Last updated on {{ last_updated_time }}</i></p>
</body>
</html>
